model/src/main/kotlin/Identifier.kt

```diff
@@ -150,5 +151,5 @@ data class Identifier(
      */
     // TODO: This is a preliminary implementation as some open questions remain, see e.g.
     // https://github.com/package-url/purl-spec/issues/33.
-    fun toPurl() = "pkg://$type/$namespace/$name@$version"
+    fun toPurl() = "pkg://$type/${namespace.urlEncode()}/${name.urlEncode()}@${version.urlEncode()}"
```

Comment:
Do you plan to also implement the other requirements from the docs? I think the most important for us are:
* type
  * "purl builders must not create invalid URLs with such double slash '//"
  * "The type is case insensitive. The canonical form is lowercase" -> should we call `toLowerCase` here?
* namespace
  * "The optional namespace contains zero or more segments, separated by slash '/'"
  * "Each namespace segment must be a percent-encoded string"
  * -> So should we actually split the namespace by "/" here and encode the segments separately?
* name
  * "The name is prefixed by a '/' separator when the namespace is not empty" -> The namespace can be empty, so the "/" before the name should be optional.
* version
  * "The version is prefixed by a '@' separator when not empty" -> The version can be empty, so the "@" should be optional.

Reply:
Not as part of this PR, nor any time soon probably. I just wanted to fix the one thing we currently clearly do wrong, and we're also affected by. Creating non-canonical purls is acceptable for now IMO, and most of the other sanity rules we currently don't violate anyway, AFAICT.